src/Routing/RouteCollection.php

```diff
@@ -104,12 +111,17 @@ public function add(Route $route, array $options = []): void
 
         // Index path prefixes (for parsing)
         $path = $route->staticPath();
-        $this->_paths[$path][] = $route;
 
         $extensions = $route->getExtensions();
         if (count($extensions) > 0) {
             $this->setExtensions($extensions);
         }
+
+        if ($path === $route->template && !$extensions) {
```

Comment:
Is the fact that the extensions check is too early for `$builder->connect()->setExtensions()` possibly grounds for concern?

Reply:
I did think about this and worse case scenario you don't get the optimizations done for static routes when extension parsing is enabled.

So for e.g. for `$builder->connect('/foo')->setExtensions(['json'])` the route will end up in the `dynamicPaths` list instead of `staticPaths` list.

Reply:
The code currently looks like `$builder->connect('/foo')->setExtensions(['json'])` would put the route in `staticPaths` because the condition of `$path === $route->template && !$extensions` will pass.  I think it all ends up working the same though because of how `parse()` has been implemented.

Reply:
Hmm, then I have to take my thumbs up back, because such a route ending up in static paths indexed without extensions, will cause it to be skipped over when the request path for the lookup has an extension.

Reply:
I will add a test for this scenario, as it will come up in application code.

Reply:
The following test does throw a `MissingRouteException` :disappointed: 

```diff
diff --git a/tests/TestCase/Routing/RouteCollectionTest.php b/tests/TestCase/Routing/RouteCollectionTest.php
index 17e5379096..ff67eec040 100644
--- a/tests/TestCase/Routing/RouteCollectionTest.php
+++ b/tests/TestCase/Routing/RouteCollectionTest.php
@@ -494,6 +494,15 @@ class RouteCollectionTest extends TestCase
         $this->assertEquals($expected, $result);
     }
 
+    public function testParseRequestExtension(): void
+    {
+        $builder = new RouteBuilder($this->collection, '/');
+        $builder->connect('/foo', ['controller' => 'Articles'])->setExtensions(['json']);
+
+        $request = new ServerRequest(['url' => '/foo.json']);
+        $this->collection->parseRequest($request);
+    }
+
     /**
      * Test parsing routes that match non-ascii urls
      */
```

Given that a route can be modified after it's added to the collection, a static route can turn into dynamic later on. So we'll either need to modify `parse()/parseRequest()` to account for that or revert this change.